File: python/linked-lists/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        for currentnode in lists:
            while currentnode:
                heapq.heappush(heap, currentnode.val)
                currentnode = currentnode.next

        head = ListNode(0)
        current = head
        while len(heap) > 0:
            current.next = ListNode(heapq.heappop(heap))
            current = current.next

        return head.next

File: c++/linked-list/merge-k-sorted-lists.cc
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        std::priority_queue<std::pair<int, ListNode*>> pq;

        ListNode* head = new ListNode();
        ListNode* pointer = head;

        for (ListNode* node : lists) {
            if (!node) {
                continue;
            }
            pq.push(std::make_pair(-node->val, node));
        }

        while (pq.size() > 0) {
            std::pair<int, ListNode*> element = pq.top();
            pq.pop();

            ListNode* next_element = element.second->next;

            pointer->next = element.second;            
            pointer = pointer->next;
            pointer->next = nullptr;

            if (next_element) {
                pq.push(std::make_pair(-next_element->val, next_element));
            }
        }

        return head->next;
    }
};
